src/Storages/Distributed/DistributedAsyncInsertBatch.cpp

```diff
@@ -245,6 +246,9 @@ void DistributedAsyncInsertBatch::sendBatch(const SettingsChanges & settings_cha
                 connection = std::move(result.front().entry);
                 compression_expected = connection->getCompression() == Protocol::Compression::Enable;
 
+                if (connection.isNull())
```

Comment:
It really seems as a logical error and ideally should be debugged. As far as I see invariant that should be satisfied is !r.entry.isNull() && r.is_usable && !(settings.distributed_insert_skip_read_only_replicas && r.is_readonly). Could you double check this and change condition to this one, please?

Reply:
Done! You're right. Because of [this line](https://github.com/ClickHouse/ClickHouse/blob/1480275ade1919651877e087fee3fda19924bc9f/src/Common/PoolWithFailoverBase.h#L303), that's the invariant that should be satisfied moving forward